### PR TITLE
Governance v2 enhancement proposal

### PR DESCRIPTION
**Pull request by CostinCarabas**

### CODE REVIEW

File: energy-integration/governance-v2/src/views.rs [hunk 1]
```diff
@@ -105,11 +103,11 @@ pub trait ViewsModule:
         proposal_id >= 1 && proposal_id <= self.proposals().len()
     }
 
-    fn proposal_reached_min_fees(&self, proposal_id: ProposalId) -> bool {
-        let accumulated_fees = self.proposals().get(proposal_id).fees.total_amount;
-        let min_fees = self.min_fee_for_propose().get();
-        accumulated_fees >= min_fees
-    }
+    // fn proposal_reached_min_fees(&self, proposal_id: ProposalId) -> bool {
```
Comment (sasurobert) [inline energy-integration/governance-v2/src/views.rs:106]: delete commented code.

File: energy-integration/governance-v2/src/views.rs [hunk 2]
```diff
@@ -52,15 +49,16 @@ pub trait ViewsModule:
...
         let total_down_veto_votes = proposal_votes.down_veto_votes;
         let third_total_votes = &total_votes / 3u64;
+        let half_total_votes = &total_votes / 2u64;
         let quorum = self.quorum().get();
 
         if total_down_veto_votes > third_total_votes {
             false
         } else {
-            total_votes >= quorum && total_up_votes > (total_down_votes + total_down_veto_votes)
+            sc_print!("total_up_votes = {}, half_total_votes = {}", total_up_votes, half_total_votes);
```
Comment (sasurobert) [inline energy-integration/governance-v2/src/views.rs:60]: no prints.

File: energy-integration/governance-v2/src/lib.rs [hunk 3]
```diff
@@ -153,110 +153,42 @@ pub trait GovernanceV2:
...
             self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
-            "Proposal is not active"
+            PROPOSAL_NOT_ACTIVE
         );
 
         let voter = self.blockchain().get_caller();
         let new_user = self.user_voted_proposals(&voter).insert(proposal_id);
         require!(new_user, ALREADY_VOTED_ERR_MSG);
 
         let user_energy = self.get_energy_amount_non_zero(&voter);
```
Comment (sasurobert) [inline energy-integration/governance-v2/src/lib.rs:163]: you need to accumulate user_energy and to compute the quorum.

Comment (sasurobert) [reply energy-integration/governance-v2/src/lib.rs:163]: total_votes should be updated by user_energgy - not by sqrt of user_energy.

File: energy-integration/governance-v2/src/proposal.rs [hunk 4]
```diff
@@ -87,5 +86,5 @@ pub struct GovernanceProposal<M: ManagedTypeApi> {
     pub proposer: ManagedAddress<M>,
     pub actions: ArrayVec<GovernanceAction<M>, MAX_GOVERNANCE_PROPOSAL_ACTIONS>,
     pub description: ManagedBuffer<M>,
-    pub fees: ProposalFees<M>,
+    pub fee_payment: EsdtTokenPayment<M>,
```
Comment (psorinionut) [inline energy-integration/governance-v2/src/proposal.rs:89]: As ProposalFees struct is no longer user, I would remove it, along with FeeEntry struct.

File: energy-integration/governance-v2/src/lib.rs [hunk 5]
```diff
@@ -153,110 +153,42 @@ pub trait GovernanceV2:
...
-        let current_block = self.blockchain().get_block_nonce();
-        self.proposal_queue_block(proposal_id).set(current_block);
-
-        self.proposal_queued_event(proposal_id, current_block);
-    }
-
-    /// Execute a previously queued proposal.
-    /// This will also clear the proposal from storage.
-    #[endpoint]
-    fn execute(&self, proposal_id: ProposalId) {
```
Comment (psorinionut) [inline energy-integration/governance-v2/src/lib.rs:217]: We no longer have the option to automatically execute proposal actions? If so, should we remove the GovernanceAction list from the GovernanceProposal (and all that comes with it -> like gas variable and checks).

Comment (sasurobert) [reply energy-integration/governance-v2/src/lib.rs:217]: we might return these - but into a separate proxy contract - on v2.

File: energy-integration/governance-v2/src/views.rs [hunk 6]
```diff
@@ -41,29 +33,63 @@ pub trait ViewsModule:
...
+
+        total_down_veto_votes > third_total_votes
+    }
+
+    fn quorum_reached(&self, proposal_id: ProposalId) -> bool {
+        let current_quorum = self.get_quorum(proposal_id);
+        let total_energy = BigUint::zero(); // TODO
+        let required_quorum = self.quorum().get();
+
+        //sc.last_global_update_week().get()
```
Comment (sasurobert) [inline energy-integration/governance-v2/src/views.rs:75]: delete unused code.

File: energy-integration/governance-v2/src/lib.rs [hunk 7]
```diff
@@ -304,17 +261,20 @@ pub trait GovernanceV2:
...
-        for fee_entry in payments.entries.iter() {
-            let payment = fee_entry.tokens;
-            self.send().direct_esdt(
-                &fee_entry.depositor_addr,
-                &payment.token_identifier,
-                payment.token_nonce,
-                &payment.amount,
-            );
-        }
+    fn refund_proposal_fee(&self, proposal_id: ProposalId, refund_percentage: u64) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply energy-integration/governance-v2/src/lib.rs:288]: Good idea - please add the remaining tokens into a list.

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — DISMISSED** [review]

**Review by psorinionut — APPROVED** [review]

File: energy-integration/governance-v2/src/views.rs [hunk 8]
```diff
@@ -41,27 +38,59 @@ pub trait ViewsModule:
...
+        let proposal = self.proposals().get(proposal_id);
+        let total_energy_for_proposal = proposal.total_energy;
+
+        let required_minimum_percentage = proposal.minimum_quorum;
+
+        let current_quorum = self.get_current_quorum(proposal_id);
+        let current_quorum_percentage =
+            current_quorum * FULL_PERCENTAGE / total_energy_for_proposal;
+
+        current_quorum_percentage > required_minimum_percentage
```
Comment (psorinionut) [inline energy-integration/governance-v2/src/views.rs:84]: Not very important, but for clarity, maybe check if `current_quorum_percentage` is greater or equal than `required_minimum_percentage`?

File: energy-integration/governance-v2/src/proposal_storage.rs [hunk 9]
```diff
@@ -59,21 +61,13 @@ impl<M: ManagedTypeApi> ProposalVotes<M> {
 pub trait ProposalStorageModule {
     fn clear_proposal(&self, proposal_id: ProposalId) {
         self.proposals().clear_entry(proposal_id);
-        self.proposal_start_block(proposal_id).clear();
-        self.proposal_queue_block(proposal_id).clear();
         self.proposal_votes(proposal_id).clear();
     }
 
     #[storage_mapper("proposals")]
```
Comment (dragos-rebegea) [inline energy-integration/governance-v2/src/proposal_storage.rs:67]: please add view

**Review by psorinionut — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**